### PR TITLE
Pull OIT into OCP pipeline

### PR DESCRIPTION
**Pull request by adammhaile**
@jupierce 
Need to wait on @smunilla  to finish builds before we can test merge or try this

### CODE REVIEW

File: jobs/build/ocp/Jenkinsfile [hunk 1]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+                            base_path = env.GITHUB_BASE_PATHS[alias]
+                            dockerfile_sub_path = val['dockerfile'].minus(base_path)
+                            dockerfile_url = "Source file: " + github_url + "/" + dockerfile_sub_path
+                        }
+                        mail(to: "${val.owners}",
+                                from: "aos-cd@redhat.com",
+                                subject: "${val.image} Dockerfile reconciliation.",
+                                body: """OIT has detected a change in the Dockerfile for ${val.image}
+                    ${dockerfile_url}
+                    This has been automatially reconciled and the new file can be seen here:
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: automatically

File: pipeline-scripts/buildlib.groovy [hunk 2]
```diff
@@ -297,13 +330,63 @@ def build_puddle(conf_url, keys, Object...args) {
...
+    int i = 0
+    int f = 0
+    for ( i = 0; i < lines.length; i++ ) {
+        fields = lines[i].split("|")
+        type = fields[0]
+        if(! result.containsKey(type)) {
+            result[type] = []
+        }
+        record = [:]
+        for ( f = 1; f = fields.length; f++ ){
```
Comment (jupierce) [inline pipeline-scripts/buildlib.groovy]: f < fields.length . This script should be testable in a local jenkins instance. Just setup a normal Jenkins pipeline and paste your code in. Prepopulate a record.log on the file system and work out the kinks.

File: pipeline-scripts/buildlib.groovy [hunk 3]
```diff
@@ -297,13 +330,63 @@ def build_puddle(conf_url, keys, Object...args) {
...
+
 def with_virtualenv(path, f) {
     final env = [
         "VIRTUAL_ENV=${path}",
         "PATH=${path}/bin:${env.PATH}",
         "PYTHON_HOME=",
     ]
     return withEnv(env, f)
+
+def parse_record_log( working_dir ) {
```
Comment (jupierce) [inline pipeline-scripts/buildlib.groovy]: Comments needed

File: pipeline-scripts/buildlib.groovy [hunk 4]
```diff
@@ -297,13 +330,63 @@ def build_puddle(conf_url, keys, Object...args) {
...
+                record[entry[0]] = entry[1]
+            }
+        }
+        result[type].add(record)
+    }
+
+    return result
+}
+
+def get_distgit_notify( record_log ) {
```
Comment (jupierce) [inline pipeline-scripts/buildlib.groovy]: Comments needed

File: jobs/build/ocp/Jenkinsfile [hunk 5]
```diff
@@ -135,6 +171,12 @@ node(TARGET_NODE) {
             buildlib.initialize()
             echo "Initializing build: #${currentBuild.number} - ${BUILD_VERSION}.?? (${BUILD_MODE})"
 
+            env.GITHUB_URLS = [:]
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: +1

File: jobs/build/ocp/Jenkinsfile [hunk 6]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
 
-            stage( "signing rpms" ) {
-                if ( SIGN_RPMS ) {
-                    sh "${env.WORKSPACE}/build-scripts/sign_rpms.sh rhaos-${BUILD_VERSION}-rhel-7-candidate openshifthosted"
-                } else {
-                    echo "RPM signing has been skipped..."
+                // OIT Method
+                OIT_WORKING = "${pwd(tmp=true)}"
+
+                stage ( "Create sources.yml" ){
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: done

File: jobs/build/ocp/Jenkinsfile [hunk 7]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
-                      PUDDLE_CONF,    // The puddle configuration file to use
-                      PUDDLE_SIGN_KEYS, // openshifthosted key
-                      "-b",   // do not fail if we are missing dependencies
-                      "-d",   // print debug information
-                      "-n",   // do not send an email for this puddle
-                      "-s",   // do not create a "latest" link since this puddle is for building images
-                      "--label=building"   // create a symlink named "building" for the puddle
-              )
-            }
+                stage( "update dist-git" ) {
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: no-push was just for initial test... would be removed of course.

File: jobs/build/ocp/Jenkinsfile [hunk 8]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+                    oit '''
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Good to know

File: jobs/build/ocp/Jenkinsfile [hunk 9]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+--working-dir ${OIT_WORKING} --group 'openshift-${BUILD_VERSION}' \
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: I miss Python...

File: jobs/build/ocp/Jenkinsfile [hunk 10]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+distgits:rebase --sources ${env.WORKSPACE}/sources.yml --version ${NEW_VERSION} \
+--release '${NEW_DOCKERFILE_RELEASE} --no-push'
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Fixed

File: jobs/build/ocp/Jenkinsfile [hunk 11]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+'''
+                }
 
-            stage( "compare dist-git" ) {
-                sh "ose_images.sh --user ocp-build compare_nodocker --branch rhaos-${BUILD_VERSION}-rhel-7 --group base"
-            }
+                record_log = buildlib.parse_record_log( OIT_WORKING )
+                distgit_notify = buildlib.get_distgit_notify( record_log )
+
+                for(distgit, val in distgit_notify) {
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Yup... was in python world mentally again.
Fixed with this: https://stackoverflow.com/a/42775560

File: jobs/build/ocp/Jenkinsfile (hunk 1 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Done

File: jobs/build/ocp/Jenkinsfile [hunk 12]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+                    http://pkgs.devel.redhat.com/cgit/${disgit}/tree/Dockerfile?id=${val.sha}
+                    """);
+                    }
+                }
 
-            stage( "update dist-git" ) {
-                sh "ose_images.sh --user ocp-build update_docker --branch rhaos-${BUILD_VERSION}-rhel-7 --group base --force --release '${NEW_DOCKERFILE_RELEASE}' --version 'v${NEW_VERSION}'"
-            }
+                stage( "build images" ) {
+                    repo_type = SIGN ? "signed" : "unsigned"
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Done

File: jobs/build/ocp/Jenkinsfile [hunk 13]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+                    sh '''
+ose_images.sh --user ocp-build build_container --branch rhaos-${BUILD_VERSION}-rhel-7 --group base \
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Done

File: jobs/build/ocp/Jenkinsfile [hunk 14]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+--repo https://raw.githubusercontent.com/openshift/aos-cd-jobs/master/build-scripts/repo-conf/aos-unsigned-building.repo"
+'''
 
-            stage( "build images" ) {
-                // TODO: Create a dynamic .repo file pointing to the exact puddle we built instead of "building" so that we can run X.Y builds in parallel
-                sh "ose_images.sh --user ocp-build build_container --branch rhaos-${BUILD_VERSION}-rhel-7 --group base --repo https://raw.githubusercontent.com/openshift/aos-cd-jobs/master/build-scripts/repo-conf/aos-unsigned-building.repo"
-            }
+                    oit '''
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Done

File: jobs/build/ocp/Jenkinsfile [hunk 15]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+--working-dir ${OIT_WORKING} --group openshift-${BUILD_VERSION} --include aos3-installation-docker \
+distgits:build-images \
+--push-to "registry.reg-aws.openshift.com:443/online" \
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Done. Don't think that was an option when I first wrote it :)

File: jobs/build/ocp/Jenkinsfile [hunk 16]
```diff
@@ -187,285 +244,348 @@ node(TARGET_NODE) {
...
+--repo_type ${repo_type} --scratcha
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: <facepalm> Not sure about that. Was just leaving scratch in for now until we can verify this more.... unless you'd rather I take it out now?

File: pipeline-scripts/buildlib.groovy [hunk 17]
```diff
@@ -48,6 +48,21 @@ def registry_login() {
     }
 }
 
+def initialize_enterprise_images_dir() {
+    VENV_DIR = "${env.WORKSPACE}/pyenv"
+    ENTERPRISE_IMAGES_DIR = "${env.WORKSPACE}/enterprise-images"
+    OIT_PATH = "${ENTERPRISE_IMAGES_DIR}/oit/oit.python"
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply pipeline-scripts/buildlib.groovy]: WTF?? I have no idea why... Atom probably got overzealous.

File: pipeline-scripts/buildlib.groovy (hunk 3 above)
Comment (adammhaile) [reply pipeline-scripts/buildlib.groovy]: Done

File: pipeline-scripts/buildlib.groovy (hunk 4 above)
Comment (adammhaile) [reply pipeline-scripts/buildlib.groovy]: Done

File: jobs/build/ocp/Jenkinsfile [hunk 18]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
-              )
-            }
+                // OIT Method
+                OIT_WORKING = "${pwd(tmp=true)}"
 
-            stage( "compare dist-git" ) {
-                sh "ose_images.sh --user ocp-build compare_nodocker --branch rhaos-${BUILD_VERSION}-rhel-7 --group base"
-            }
+                stage( "update dist-git" ) {
+                  sh "echo 'ose: ${env.OSE_DIR}' > ${env.WORKSPACE}/sources.yml"
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: Can we just use a writeFile?

File: jobs/build/ocp/Jenkinsfile [hunk 19]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+                            github_url = github_url.replaceFirst(":", "/")
+                            base_path = env.GITHUB_BASE_PATHS[alias]
+                            dockerfile_sub_path = val['dockerfile'].minus(base_path)
+                            dockerfile_url = "Source file: " + github_url + "/" + dockerfile_sub_path
+                        }
+                        mail(to: "${val.owners}",
+                                from: "aos-cd@redhat.com",
+                                subject: "${val.image} Dockerfile reconciliation.",
+                                body: """OIT has detected a change in the Dockerfile for ${val.image}
+                    ${dockerfile_url}
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: """ will maintain this indentation whitepace in the email, which won't look right.

File: jobs/build/ocp/Jenkinsfile [hunk 20]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
-                // TODO: Create a dynamic .repo file pointing to the exact puddle we built instead of "building" so that we can run X.Y builds in parallel
-                sh "ose_images.sh --user ocp-build build_container --branch rhaos-${BUILD_VERSION}-rhel-7 --group base --repo https://raw.githubusercontent.com/openshift/aos-cd-jobs/master/build-scripts/repo-conf/aos-unsigned-building.repo"
-            }
+                record_log = buildlib.parse_record_log( OIT_WORKING )
+                distgit_notify = buildlib.get_distgit_notify( record_log )
+
+                for(kv in buildlib.mapToList(distgit_notify)) {
+                    distgit = kv[0]
+                    val = kv[1]
+                    if(val.owners) {
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: CD needs an email even if there are no owners specified.

File: jobs/build/ocp/Jenkinsfile [hunk 21]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+                }
 
-            stage( "build images" ) {
-                // TODO: Create a dynamic .repo file pointing to the exact puddle we built instead of "building" so that we can run X.Y builds in parallel
-                sh "ose_images.sh --user ocp-build build_container --branch rhaos-${BUILD_VERSION}-rhel-7 --group base --repo https://raw.githubusercontent.com/openshift/aos-cd-jobs/master/build-scripts/repo-conf/aos-unsigned-building.repo"
-            }
+                record_log = buildlib.parse_record_log( OIT_WORKING )
+                distgit_notify = buildlib.get_distgit_notify( record_log )
+
+                for(kv in buildlib.mapToList(distgit_notify)) {
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: Comments needed.

File: jobs/build/ocp/Jenkinsfile [hunk 22]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+                    distgit = kv[0]
+                    val = kv[1]
+                    if(val.owners) {
+                        alias = val['source_alias']
+                        dockerfile_url = ""
+                        if ( alias ){
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: When would we expect this to be missing?

File: jobs/build/ocp/Jenkinsfile [hunk 23]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+                    http://pkgs.devel.redhat.com/cgit/${disgit}/tree/Dockerfile?id=${val.sha}
+                    """);
+                    }
+                }
 
-            if ( EARLY_LATEST_HACK.toBoolean() ) {
+                stage( "build images" ) {
+                    sh """
+ose_images.sh --user ocp-build build_container --branch rhaos-${BUILD_VERSION}-rhel-7 --group base \\
+--repo https://raw.githubusercontent.com/openshift/aos-cd-jobs/master/build-scripts/repo-conf/aos-unsigned-building.repo"
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: extra double quote

File: jobs/build/ocp/Jenkinsfile [hunk 24]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+"""
+
+                    oit """
+--working-dir ${OIT_WORKING} --group openshift-${BUILD_VERSION} --include aos3-installation-docker \\
+distgits:build-images \\
+--push-to-defaults --repo_type unsigned --scratch
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: Why is this scratch?

File: jobs/build/ocp/Jenkinsfile [hunk 25]
```diff
@@ -187,285 +242,346 @@ node(TARGET_NODE) {
...
+                        OA_CHANGELOG = buildlib.read_changelog( "openshift-ansible.spec" )
                     }
-                    OA_CHANGELOG = buildlib.read_changelog( "openshift-ansible.spec" )
                 }
-            }
 
-            if ( IS_TEST_MODE ) {
-                error( "This is as far as the test process can proceed without triggering builds" )
-            }
+                if ( IS_TEST_MODE ) {
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: That happened when I rebased in the latest I believe... not really sure why.

File: jobs/build/ocp/Jenkinsfile (hunk 18 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Still learning groovy... I will change that.

File: jobs/build/ocp/Jenkinsfile (hunk 23 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: fixed

File: jobs/build/ocp/Jenkinsfile (hunk 24 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: I asked you about that in a previous comment... I was just being safe for now.

File: jobs/build/ocp/Jenkinsfile (hunk 22 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: I thought some of the config.yml files had no source alias.

Comment (jupierce) [reply jobs/build/ocp/Jenkinsfile]: If they don't have an alias, they can't be rebased and there will be no notify entry.

**Comment by adammhaile** [comment]
Fair enough.

File: jobs/build/ocp/Jenkinsfile (hunk 24 above)
Comment (jupierce) [reply jobs/build/ocp/Jenkinsfile]: --scatch means the image will be temporary and with an unpredicable name/tag. We can't use it for the build process.

**Comment by adammhaile** [comment]
buildlib code tested in jenkins instance and updated here
working on refresh images

**Comment by adammhaile** [comment]
@jupierce now with refresh-images

File: jobs/build/ocp/Jenkinsfile [hunk 26]
```diff
@@ -452,30 +493,80 @@ node(TARGET_NODE) {
...
+--working-dir ${OIT_WORKING} --group 'openshift-${BUILD_VERSION}' \\
+distgits:rebase --sources ${env.WORKSPACE}/sources.yml --version ${NEW_VERSION} \\
+--release ${NEW_DOCKERFILE_RELEASE}
+"""
+            }
+
+            record_log = buildlib.parse_record_log( OIT_WORKING )
+            distgit_notify = buildlib.get_distgit_notify( record_log )
+
+            for(kv in buildlib.mapToList(distgit_notify)) {
```
Comment (jupierce) [inline jobs/build/ocp/Jenkinsfile]: Needs comments. Does this 'for in' actually run in Jenkins?

File: jobs/build/refresh-images/Jenkinsfile [hunk 27]
```diff
@@ -78,12 +87,30 @@ node('openshift-build-1') {
...
+
+                // ose_images method
                 sh "ose_images.sh --user ocp-build update_docker ${rhel_arg} --version ${VERSION_OVERRIDE} --release ${RELEASE_OVERRIDE} --force --branch rhaos-${OSE_MAJOR}.${OSE_MINOR}-rhel-7 --group ${OSE_GROUP}"
                 sh "ose_images.sh --user ocp-build build --branch rhaos-${OSE_MAJOR}.${OSE_MINOR}-rhel-7 --group ${OSE_GROUP} --repo ${OSE_REPO}"
                 sh "sudo env \"PATH=${env.PATH}\" ose_images.sh push --branch rhaos-${OSE_MAJOR}.${OSE_MINOR}-rhel-7 --group ${OSE_GROUP}"
+                // end ose_images method
+
+                // OIT method
+                buildlib.write_sources_file()
+                buildlib.oit """
```
Comment (jupierce) [inline jobs/build/refresh-images/Jenkinsfile]: Should be no rebase during refresh. Look at what ose_images is doing.

File: pipeline-scripts/buildlib.groovy [hunk 28]
```diff
@@ -306,4 +341,79 @@ def with_virtualenv(path, f) {
     return withEnv(env, f)
 }
 
-return this
+// Parse record.log from OIT into a map
```
Comment (jupierce) [inline pipeline-scripts/buildlib.groovy]: What does the map look like (I know I can just read the code, but save someone time)?

File: pipeline-scripts/buildlib.groovy [hunk 29]
```diff
@@ -306,4 +341,79 @@ def with_virtualenv(path, f) {
...
+            }
+        }
+        result[type].add(record)
+    }
+
+    return result
+}
+
+
+// gets map of emails to notify from output of parse_record_log
```
Comment (jupierce) [inline pipeline-scripts/buildlib.groovy]: What does the map look like?

File: jobs/build/ocp/Jenkinsfile (hunk 26 above)
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Pulled from SO questions about Jenkins/Groovy, but easy enough to make it a normal for loop which is obviously safer.

File: jobs/build/ocp/Jenkinsfile [hunk 30]
```diff
@@ -452,30 +493,80 @@ node(TARGET_NODE) {
...
+                alias = val['source_alias']
+                dockerfile_url = ""
+                github_url = env.GITHUB_URLS[alias]
+                github_url = github_url.minus(".git")
+                github_url = github_url.minus("git@")
+                github_url = github_url.replaceFirst(":", "/")
+                base_path = env.GITHUB_BASE_PATHS[alias]
+                dockerfile_sub_path = val['dockerfile'].minus(base_path)
+                dockerfile_url = "Source file: " + github_url + "/" + dockerfile_sub_path
+                mail(to: "aos-cd@redhat.com,${val.owners}",
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply jobs/build/ocp/Jenkinsfile]: Nothing magic about it, that value flows down from here: https://github.com/openshift/enterprise-images/blob/master/oit/packages/image.py#L257
Where it's already delimited. If no owners, that value will be a blank string and we'll have a trailing comma and at least in testing on my local jenkins instance it didn't complain about that.

File: jobs/build/refresh-images/Jenkinsfile (hunk 27 above)
Comment (adammhaile) [reply jobs/build/refresh-images/Jenkinsfile]: Fixed.

File: jobs/build/ocp/Jenkinsfile (hunk 26 above)
Comment (jupierce) [reply jobs/build/ocp/Jenkinsfile]: I'm fine with it if it works. I've just had trouble with 'in' with Jenkins.

**Comment by adammhaile** [comment]
@jupierce most recent issues addressed.

**Comment by adammhaile** [comment]
Fair enough, I'm all for the least troublesome method. Now when will
Jenkins support python directly? :P

**Comment by stevekuznetsov** [comment]
🔥 🔥 🔥

**Pull request merged**